examples/ul2/src/mod_print_callback.py

```diff
@@ -0,0 +1,214 @@
+        assert self.print_frequency >= 1
+        self._steps_since_last_print = print_frequency  # This will cause it to print on the first batch, will be reset to 0 each print.
+
+        self.max_length = int(max_length)
+        assert self.max_length >= 1
+
+        self._raise_on_error = bool(raise_on_failure)
+
+        self._generate_kwargs = generate_kwargs if generate_kwargs else {}
+
+        self._verified_collator = False
+        self._decoder_only_format = None
+
+    def init(self, state: State, logger: Logger) -> None:
+        """Verifies model class."""
+        del logger  # unused
+
+        assert hasattr(
+            state.model, 'model'
+        ), 'This callback requires that state.model is a HuggingFaceWrapper instance.'
+        hf_model = state.model.model
+        assert isinstance(
+            hf_model, PreTrainedModel
+        ), 'This class only works for models that subclass `transformers.PreTrainedModel`.'
+
+    def after_dataloader(self, state: State, logger: Logger) -> None:
+        """Periodically prints formatted outputs generated by the model."""
+        del logger  # unused
+
+        if dist.get_global_rank() != 0:
+            return
+
+        if not self._verified_collator:
+            assert isinstance(
+                state.train_dataloader.collate_fn, MixtureOfDenoisersCollator
+            ), 'The dataloader must use the `MixtureOfDenoisersCollator` as the collate function.'
+            self._verified_collator = True
+            self._decoder_only_format = bool(
+                state.train_dataloader.collate_fn.decoder_only_format)
+
+        self._steps_since_last_print += 1
+        if self._steps_since_last_print < self.print_frequency:
+            return
+
+        if state.is_model_ddp:
+            model = state.model.module.model  # Extract from DDP wrapping
+        else:
+            model = state.model.model
+
+        input_str, target_str, output_str = self._generate_input_target_output_strings(
+            state.batch, model)
+
+        try:
+            example_string = self._format_strings(input_str, target_str,
+                                                  output_str)
+            print(
+                f'\n\n{self.__class__.__name__} printout...\n{example_string}\n',
+                file=sys.stdout,
+                flush=True)
+
+        except AssertionError:
+            if self._raise_on_error:
+                raise
+            print(
+                f'\n\n{self.__class__.__name__}: Failed to parse input/target/output strings. This is expected if the model is early in training. Set `raise_on_failure=True` if you want this to failure generate an error.\n',
+                file=sys.stdout,
+                flush=True)
+
+        self._steps_since_last_print = 0
+
+    def _generate_input_target_output_strings(
+            self, batch: Dict[str, torch.TensorType],
+            model: PreTrainedModel) -> List[str]:
```

Comment:
seems like a prime candidate to be a function (or staticmethod) rather than method